components.go

```diff
@@ -0,0 +1,133 @@
+package discordgo
+
+import (
+	"encoding/json"
+)
+
+// ComponentType is type of component.
+type ComponentType uint
+
+// MessageComponent types.
+const (
+	ActionsRowComponent ComponentType = 1
+	ButtonComponent     ComponentType = 2
+)
+
+// MessageComponent is a base interface for all message components.
+type MessageComponent interface {
+	json.Marshaler
+	Type() ComponentType
+}
+
+type unmarshalableMessageComponent struct {
+	MessageComponent
+}
+
+// UnmarshalJSON is a helper function to unmarshal MessageComponent object.
+func (umc *unmarshalableMessageComponent) UnmarshalJSON(src []byte) (err error) {
+	var v struct {
+		Type ComponentType `json:"type"`
+	}
+	err = json.Unmarshal(src, &v)
+	if err != nil {
+		return
+	}
+
+	var data MessageComponent
+	switch v.Type {
+	case ActionsRowComponent:
+		v := ActionsRow{}
+		err = json.Unmarshal(src, &v)
```

Comment:
I'm getting the following panic in this example:
```go
s, _ := discordgo.New("Bot <token>")

ms := &discordgo.MessageSend{
	Content: "test",
	Components: []discordgo.MessageComponent{
		&discordgo.ActionsRow{
                    Components: []discordgo.MessageComponent{
			&discordgo.Button{
				Label:    "test",
				CustomID: "test",
			},
		    },
		},
	},
}
m, err := s.ChannelMessageSendComplex("<id>", ms)
if err != nil {
	panic(err)
}
```
```go
panic: json unmarshal: json: cannot unmarshal object into Go struct field .components of type discordgo.MessageComponent
```
Are you sure that we don't need to have a custom `UnmarshalJSON` function for `ActionsRow`, since it contains a slice of `MessageComponent`, which is an interface? Did you see this type of issue in your testing?

Reply:
Nope, thanks for pointing it out. However, on my opinion would be more cleaner if we just make for the slice another type, like I was doing previously.

Reply:
I'd much rather have the tedious work done once internally than have the embedding as a confusing API surface to newcomers (in addition to, say, having to manually convert each element of the slice if they wanted to reuse the components in another message). Having the separate type just isn't something that should be exposed by the library if we can help it.

Reply:
Yeah, sure.